tests/tracking/test_context.py

```diff
@@ -0,0 +1,223 @@
+
+
+def test_context_provider_registry_register_entrypoints():
+    provider_class = mock.Mock()
+    mock_entrypoint = mock.Mock()
+    mock_entrypoint.load.return_value = provider_class
+
+    with mock.patch(
+        "entrypoints.get_group_all", return_value=[mock_entrypoint]
+    ) as mock_get_group_all:
+        registry = ContextProviderRegistry()
+        registry.register_entrypoints()
+
+    assert set(registry) == {provider_class.return_value}
+    mock_entrypoint.load.assert_called_once_with()
+    mock_get_group_all.assert_called_once_with("mlflow.context_provider")
+
+
+@pytest.mark.parametrize("exception",
+                         [AttributeError("test exception"), ImportError("test exception")])
+def test_context_provider_registry_register_entrypoints_handles_exception(exception):
+    mock_entrypoint = mock.Mock()
+    mock_entrypoint.load.side_effect = exception
+
+    with mock.patch(
+        "entrypoints.get_group_all", return_value=[mock_entrypoint]
+    ) as mock_get_group_all:
+        registry = ContextProviderRegistry()
+        # Check that the raised warning contains the message from the original exception
+        with pytest.warns(UserWarning, match="test exception"):
+            registry.register_entrypoints()
+
+    mock_entrypoint.load.assert_called_once_with()
+    mock_get_group_all.assert_called_once_with("mlflow.context_provider")
+
+
+def _currently_registered_context_provider_classes():
+    return {
+        provider.__class__
+        for provider in mlflow.tracking.context._context_provider_registry
+    }
+
+
+def test_registry_instance_defaults():
+    expected_context_provider_classes = {DefaultContext, GitContext, DatabricksNotebookContext}
+    assert expected_context_provider_classes.issubset(
+        _currently_registered_context_provider_classes()
+    )
+
+
+def test_registry_instance_loads_entrypoints():
+
+    class MockContext(object):
+        pass
+
+    mock_entrypoint = mock.Mock()
+    mock_entrypoint.load.return_value = MockContext
+
+    with mock.patch(
+        "entrypoints.get_group_all", return_value=[mock_entrypoint]
+    ) as mock_get_group_all:
+        # Entrypoints are registered at import time, so we need to reload the module to register th
+        # entrypoint given by the mocked extrypoints.get_group_all
+        reload(mlflow.tracking.context)
+
+    assert MockContext in _currently_registered_context_provider_classes()
+    mock_get_group_all.assert_called_once_with("mlflow.context_provider")
+
+
+@pytest.mark.large
+def test_context_provider_registry_with_installed_plugin(tmp_wkdir):
+    """This test requires the package in tests/resources/mlflow-test-plugin to be installed"""
+
+    reload(mlflow.tracking.context)
+
+    from mlflow_test_plugin import PluginContextProvider
+    assert PluginContextProvider in _currently_registered_context_provider_classes()
+
+    with mock.patch.object(PluginContextProvider, "tags", return_value={"test": "tag"}):
```

Comment:
Maybe we could just have the `PluginContextProvider` return this, to avoid the need for mocking it?

Reply:
Potentially - I did it this way so that the expected values asserted by the test are not dependent on some unseen implementation, but if you prefer the no-mock solution, I'm happy to change.

Reply:
It's a minor thing -- I was considering this particular test case as more of an "integration test", where it's nice to avoid mocks in case they hide some subtle broken behavior.

Reply:
Fair comment; I'll make the change.

Reply:
I remembered an additional reason why I did it this way - the fluent tests would break if the test plugin returns tags, as they assert the set of tags that should be generated by default. I've tweaked the test to so the plugin always returns tags and the mock just 'switches it on', and added a comment to explain this - let me know what you think.

Reply:
Ahh, interesting -- this is fine as you've updated it.